File: archaeologist-game/js/dig-game.js
// Мини-игра "Раскопки" (Три в ряд)

class DigGame {
    constructor(canvasId, onComplete, onFail) {
        this.canvas = document.getElementById(canvasId);
        this.ctx = this.canvas.getContext('2d');
        this.onComplete = onComplete;
        this.onFail = onFail;
        
        // Настройки игры
        this.gridSize = 35;
        this.cols = 8;
        this.rows = 8;
        
        // Состояние игры
        this.grid = [];
        this.selectedCell = null;
        this.score = 0;
        this.targetScore = 500;
        this.moves = 20;
        this.isPlaying = false;
        this.isAnimating = false;
        this.animationId = null;
        
        // Инициализация размеров canvas
        this.resize();
        window.addEventListener('resize', () => this.resize());
        
        // Обработчики событий
        this.setupControls();
    }
    
    resize() {
        const maxWidth = Math.min(window.innerWidth - 40, 600);
        const maxHeight = Math.min(window.innerHeight - 200, 600);
        
        this.canvas.width = maxWidth;
        this.canvas.height = maxHeight;
        
        this.gridSize = Math.floor(Math.min(
            this.canvas.width / this.cols,
            this.canvas.height / this.rows
        )) - 2;
    }
    
    setupControls() {
        this.canvas.addEventListener('click', (e) => this.handleClick(e));
        this.canvas.addEventListener('touchstart', (e) => {
            e.preventDefault();
            const touch = e.touches[0];
            const rect = this.canvas.getBoundingClientRect();
            this.handleClick({
                clientX: touch.clientX,
                clientY: touch.clientY
            });
        });
    }
    
    start(level = 1) {
        this.level = level;
        this.score = 0;
        this.moves = 20 + level * 2;
        this.targetScore = 500 + level * 100;
        this.selectedCell = null;
        this.isPlaying = true;
        this.isAnimating = false;
        
        // Генерация сетки
        this.generateGrid();
        
        // Проверка на готовые совпадения и их удаление
        this.removeMatches(false);
        
        // Обновление UI
        this.updateUI();
        
        // Запуск отрисовки
        this.draw();
    }
    
    generateGrid() {
        this.grid = [];
        for (let row = 0; row < this.rows; row++) {
            this.grid[row] = [];
            for (let col = 0; col < this.cols; col++) {
                this.grid[row][col] = this.randomElement();
            }
        }
    }
    
    randomElement() {
        const index = Math.floor(Math.random() * gameElements.length);
        return gameElements[index];
    }
    
    handleClick(e) {
        if (!this.isPlaying || this.isAnimating) return;
        
        const rect = this.canvas.getBoundingClientRect();
        const scaleX = this.canvas.width / rect.width;
        const scaleY = this.canvas.height / rect.height;
        
        const x = (e.clientX - rect.left) * scaleX;
        const y = (e.clientY - rect.top) * scaleY;
        
        const col = Math.floor(x / (this.gridSize + 2));
        const row = Math.floor(y / (this.gridSize + 2));
        
        if (row >= 0 && row < this.rows && col >= 0 && col < this.cols) {
            this.selectCell(row, col);
        }
    }
    
    selectCell(row, col) {
        if (!this.selectedCell) {
            // Первая ячейка выбрана
            this.selectedCell = { row, col };
        } else {
            // Вторая ячейка выбрана - пробуем поменять
            const dr = Math.abs(row - this.selectedCell.row);
            const dc = Math.abs(col - this.selectedCell.col);
            
            // Проверяем, что ячейки соседние
            if ((dr === 1 && dc === 0) || (dr === 0 && dc === 1)) {
                this.swapCells(this.selectedCell.row, this.selectedCell.col, row, col);
            }
            this.selectedCell = null;
        }
        
        this.draw();
    }
    
    async swapCells(row1, col1, row2, col2) {
        this.isAnimating = true;
        
        // Меняем местами
        const temp = this.grid[row1][col1];
        this.grid[row1][col1] = this.grid[row2][col2];
        this.grid[row2][col2] = temp;
        
        this.draw();
        
        // Проверяем на совпадения
        await new Promise(resolve => setTimeout(resolve, 300));
        
        const matches = this.findMatches();
        if (matches.length > 0) {
            // Есть совпадения - удаляем
            this.moves--;
            await this.removeMatches(true);
        } else {
            // Нет совпадений - возвращаем обратно
            const temp = this.grid[row1][col1];
            this.grid[row1][col1] = this.grid[row2][col2];
            this.grid[row2][col2] = temp;
            this.draw();
        }
        
        this.isAnimating = false;
        this.updateUI();
        this.checkGameState();
    }
    
    findMatches() {
        const matches = [];
        
        // Горизонтальные совпадения
        for (let row = 0; row < this.rows; row++) {
            for (let col = 0; col < this.cols - 2; col++) {
                const type = this.grid[row][col].type;
                let matchLength = 1;
                
                while (col + matchLength < this.cols && 
                       this.grid[row][col + matchLength].type === type) {
                    matchLength++;
                }
                
                if (matchLength >= 3) {
                    for (let i = 0; i < matchLength; i++) {
                        matches.push({ row, col: col + i });
                    }
                }
            }
        }
        
        // Вертикальные совпадения
        for (let col = 0; col < this.cols; col++) {
            for (let row = 0; row < this.rows - 2; row++) {
                const type = this.grid[row][col].type;
                let matchLength = 1;
                
                while (row + matchLength < this.rows && 
                       this.grid[row + matchLength][col].type === type) {
                    matchLength++;
                }
                
                if (matchLength >= 3) {
                    for (let i = 0; i < matchLength; i++) {
                        matches.push({ row: row + i, col });
                    }
                }
            }
        }
        
        // Удаляем дубликаты
        const unique = [];
        const seen = new Set();
        matches.forEach(m => {
            const key = `${m.row},${m.col}`;
            if (!seen.has(key)) {
                seen.add(key);
                unique.push(m);
            }
        });
        
        return unique;
    }
    
    async removeMatches(addScore) {
        let hasMatches = true;
        
        while (hasMatches) {
            const matches = this.findMatches();
            
            if (matches.length === 0) {
                hasMatches = false;
                break;
            }
            
            // Начисление очков
            if (addScore) {
                const matchCount = matches.length;
                let points = 0;
                
                if (matchCount >= 5) points = 35;
                else if (matchCount >= 4) points = 25;
                else if (matchCount >= 3) points = 10;
                
                this.score += points * matchCount / 3;
            }
            
            // Удаляем совпадения
            matches.forEach(m => {
                this.grid[m.row][m.col] = null;
            });
            
            this.draw();
            await new Promise(resolve => setTimeout(resolve, 200));
            
            // Падение элементов
            this.dropElements();
            this.draw();
            await new Promise(resolve => setTimeout(resolve, 300));
            
            // Заполнение пустых мест
            this.fillEmpty();
            this.draw();
            await new Promise(resolve => setTimeout(resolve, 300));
        }
        
        this.updateUI();
    }
    
    dropElements() {
        for (let col = 0; col < this.cols; col++) {
            let writeRow = this.rows - 1;
            
            for (let row = this.rows - 1; row >= 0; row--) {
                if (this.grid[row][col] !== null) {
                    this.grid[writeRow][col] = this.grid[row][col];
                    if (writeRow !== row) {
                        this.grid[row][col] = null;
                    }
                    writeRow--;
                }
            }
        }
    }
    
    fillEmpty() {
        for (let row = 0; row < this.rows; row++) {
            for (let col = 0; col < this.cols; col++) {
                if (this.grid[row][col] === null) {
                    this.grid[row][col] = this.randomElement();
                }
            }
        }
    }
    
    draw() {
        // Очистка
        this.ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
        
        // Рисуем фон
        this.ctx.fillStyle = '#f5f5dc';
        this.ctx.fillRect(0, 0, this.canvas.width, this.canvas.height);
        
        // Рисуем сетку
        for (let row = 0; row < this.rows; row++) {
            for (let col = 0; col < this.cols; col++) {
                const x = col * (this.gridSize + 2);
                const y = row * (this.gridSize + 2);
                
                // Фон ячейки
                this.ctx.fillStyle = (row + col) % 2 === 0 ? '#fff' : '#f0f0f0';
                this.ctx.fillRect(x, y, this.gridSize, this.gridSize);
                
                // Выделение выбранной ячейки
                if (this.selectedCell && 
                    this.selectedCell.row === row && 
                    this.selectedCell.col === col) {
                    this.ctx.strokeStyle = '#667eea';
                    this.ctx.lineWidth = 3;
                    this.ctx.strokeRect(x + 2, y + 2, this.gridSize - 4, this.gridSize - 4);
                }
                
                // Рисуем элемент
                if (this.grid[row][col]) {
                    this.drawEmoji(
                        this.grid[row][col].icon,
                        x + this.gridSize / 2,
                        y + this.gridSize / 2,
                        this.gridSize * 0.7
                    );
                }
            }
        }
    }
    
    drawEmoji(emoji, x, y, size) {
        this.ctx.font = `${size}px Arial`;
        this.ctx.textAlign = 'center';
        this.ctx.textBaseline = 'middle';
        this.ctx.fillText(emoji, x, y);
    }
    
    updateUI() {
        document.getElementById('dig-score').textContent = `Очки: ${Math.floor(this.score)}`;
        document.getElementById('dig-target').textContent = `Цель: ${this.targetScore}`;
        document.getElementById('dig-moves').textContent = `Ходы: ${this.moves}`;
    }
    
    checkGameState() {
        if (this.score >= this.targetScore) {
            this.win();
        } else if (this.moves <= 0) {
            this.lose();
        }
    }
    
    win() {
        this.isPlaying = false;
        this.showMessage('🎉 Отлично! Цель достигнута!', 'success');
        
        setTimeout(() => {
            this.onComplete(Math.floor(this.score));
        }, 1500);
    }
    
    lose() {
        this.isPlaying = false;
        this.showMessage('😔 Не хватило ходов!', 'error');
        
        setTimeout(() => {
            this.onFail();
        }, 1500);
    }
    
    showMessage(text, type) {
        const messageEl = document.getElementById('dig-message');
        messageEl.textContent = text;
        messageEl.style.display = 'block';
        messageEl.style.color = type === 'success' ? '#27ae60' : '#e74c3c';
        
        setTimeout(() => {
            messageEl.style.display = 'none';
        }, 1500);
    }
}

File: src/main/java/ru/des88/dinopedia/DigGame.java
package ru.des88.dinopedia;

import org.teavm.jso.browser.Window;
import org.teavm.jso.canvas.CanvasRenderingContext2D;
import org.teavm.jso.dom.html.HTMLCanvasElement;
import org.teavm.jso.dom.html.HTMLElement;
import org.teavm.jso.dom.events.MouseEvent;
import org.teavm.jso.dom.events.TouchEvent;
import org.teavm.jso.dom.events.Touch;
import org.teavm.jso.dom.events.EventListener;
import org.teavm.jso.core.JSArrayReader;
import ru.des88.dinopedia.core.UIHelper;
import ru.des88.dinopedia.data.GameElement;

/**
 * Мини-игра «Раскопки» — головоломка «Три в ряд».
 * Поля: 9×9 (уровни 1–3), 12×12 (4–6), 16×16 (7+).
 * Элементы: кости динозавра (целевые), камни, кирпичи, кусты.
 * Цель: набрать 500–1000 очков, собирая ряды из 3+ элементов.
 */
public class DigGame {
    private final HTMLCanvasElement canvas;
    private final CanvasRenderingContext2D ctx;
    private final PathGame.IntCallback onCompleteCallback;
    private final Runnable onFailCallback;

    private int gridSize = 9;
    private int cellSize = 40;
    private int padding = 4;

    private GameElement[][] grid;
    private int selectedRow = -1;
    private int selectedCol = -1;
    private int score = 0;
    private int targetScore = 500;
    private int moves = 25;
    private int bonesCollected = 0;
    private int level = 1;
    private boolean playing = false;
    private boolean animating = false;

    public DigGame(HTMLCanvasElement canvas, PathGame.IntCallback onComplete, Runnable onFail) {
        this.canvas = canvas;
        this.ctx = (CanvasRenderingContext2D) canvas.getContext("2d");
        this.onCompleteCallback = onComplete;
        this.onFailCallback = onFail;
        setupInput();
    }

    // === START ===

    public void start(int lvl) {
        level = lvl;
        if (lvl <= 3) gridSize = 9;
        else if (lvl <= 6) gridSize = 12;
        else gridSize = 16;

        targetScore = Math.min(1000, 500 + level * 50);
        moves = 20 + level * 2;
        score = 0;
        bonesCollected = 0;
        selectedRow = -1;
        selectedCol = -1;
        playing = true;
        animating = false;

        resize();
        generateGrid();
        removeInitialMatches();
        updateUI();
        draw();
    }

    private void resize() {
        int maxDim = Math.min(Math.min(UIHelper.windowWidth() - 40, UIHelper.windowHeight() - 180), 700);
        cellSize = Math.max(24, Math.min(50, (maxDim - padding * (gridSize + 1)) / gridSize));
        int dim = cellSize * gridSize + padding * (gridSize + 1);
        canvas.setWidth(dim);
        canvas.setHeight(dim);
    }

    // === GRID ===

    private void generateGrid() {
        grid = new GameElement[gridSize][gridSize];
        for (int r = 0; r < gridSize; r++)
            for (int c = 0; c < gridSize; c++)
                grid[r][c] = GameElement.random();
    }

    private void removeInitialMatches() {
        int safety = 100;
        while (findMatches().length > 0 && safety-- > 0) {
            int[][] matches = findMatches();
            for (int[] m : matches) {
                grid[m[0]][m[1]] = GameElement.random();
            }
        }
    }

    // === INPUT ===

    private void setupInput() {
        canvas.onClick(e -> handleClick(e.getClientX(), e.getClientY()));
        canvas.addEventListener("touchstart", (EventListener<TouchEvent>) e -> {
            e.preventDefault();
            JSArrayReader<Touch> touches = e.getChangedTouches();
            if (touches.getLength() > 0) {
                Touch t = touches.get(0);
                handleClick(t.getClientX(), t.getClientY());
            }
        });
    }

    private void handleClick(double clientX, double clientY) {
        if (!playing || animating) return;
        double rect = canvas.getBoundingClientRect().getLeft();
        double rectTop = canvas.getBoundingClientRect().getTop();
        double scaleX = canvas.getWidth() / (double) canvas.getBoundingClientRect().getWidth();
        double scaleY = canvas.getHeight() / (double) canvas.getBoundingClientRect().getHeight();
        double x = (clientX - rect) * scaleX;
        double y = (clientY - rectTop) * scaleY;

        int total = cellSize + padding;
        int col = (int)((x - padding) / total);
        int row = (int)((y - padding) / total);
        if (row < 0 || row >= gridSize || col < 0 || col >= gridSize) return;

        selectCell(row, col);
    }

    private void selectCell(int row, int col) {
        if (selectedRow < 0) {
            selectedRow = row;
            selectedCol = col;
        } else {
            int dr = Math.abs(row - selectedRow);
            int dc = Math.abs(col - selectedCol);
            if ((dr == 1 && dc == 0) || (dr == 0 && dc == 1)) {
                trySwap(selectedRow, selectedCol, row, col);
            }
            selectedRow = -1;
            selectedCol = -1;
        }
        draw();
    }

    // === GAME LOGIC ===

    private void trySwap(int r1, int c1, int r2, int c2) {
        animating = true;
        GameElement tmp = grid[r1][c1];
        grid[r1][c1] = grid[r2][c2];
        grid[r2][c2] = tmp;
        draw();
        Window.setTimeout(() -> {
            int[][] matches = findMatches();
            if (matches.length > 0) {
                moves--;
                processMatches();
            } else {
                // Swap back
                GameElement t2 = grid[r1][c1];
                grid[r1][c1] = grid[r2][c2];
                grid[r2][c2] = t2;
                draw();
                animating = false;
                updateUI();
            }
        }, 200);
    }

    private int[][] findMatches() {
        boolean[][] matched = new boolean[gridSize][gridSize];
        int count = 0;

        // Horizontal
        for (int r = 0; r < gridSize; r++) {
            int c = 0;
            while (c < gridSize - 2) {
                String type = grid[r][c].type;
                int len = 1;
                while (c + len < gridSize && grid[r][c + len].type.equals(type)) len++;
                if (len >= 3) {
                    for (int i = 0; i < len; i++) {
                        if (!matched[r][c + i]) { matched[r][c + i] = true; count++; }
                    }
                }
                c += len;
            }
        }
        // Vertical
        for (int c = 0; c < gridSize; c++) {
            int r = 0;
            while (r < gridSize - 2) {
                String type = grid[r][c].type;
                int len = 1;
                while (r + len < gridSize && grid[r + len][c].type.equals(type)) len++;
                if (len >= 3) {
                    for (int i = 0; i < len; i++) {
                        if (!matched[r + i][c]) { matched[r + i][c] = true; count++; }
                    }
                }
                r += len;
            }
        }

        int[][] result = new int[count][2];
        int idx = 0;
        for (int r = 0; r < gridSize; r++)
            for (int c = 0; c < gridSize; c++)
                if (matched[r][c]) { result[idx][0] = r; result[idx][1] = c; idx++; }
        return result;
    }

    private void processMatches() {
        int[][] matches = findMatches();
        if (matches.length == 0) {
            animating = false;
            updateUI();
            checkState();
            return;
        }

        int mc = matches.length;
        int points;
        if (mc >= 5) points = 35;
        else if (mc >= 4) points = 25;
        else points = 10;

        int boneCount = 0;
        for (int[] m : matches) {
            if (grid[m[0]][m[1]].isTarget) boneCount++;
        }
        score += points * (mc / 3) + boneCount * 5;
        bonesCollected += boneCount;

        for (int[] m : matches) grid[m[0]][m[1]] = null;
        draw();

        Window.setTimeout(() -> {
            dropElements();
            draw();
            Window.setTimeout(() -> {
                fillEmpty();
                draw();
                Window.setTimeout(() -> processMatches(), 200);
            }, 200);
        }, 250);
    }

    private void dropElements() {
        for (int c = 0; c < gridSize; c++) {
            int writeRow = gridSize - 1;
            for (int r = gridSize - 1; r >= 0; r--) {
                if (grid[r][c] != null) {
                    grid[writeRow][c] = grid[r][c];
                    if (writeRow != r) grid[r][c] = null;
                    writeRow--;
                }
            }
        }
    }

    private void fillEmpty() {
        for (int r = 0; r < gridSize; r++)
            for (int c = 0; c < gridSize; c++)
                if (grid[r][c] == null) grid[r][c] = GameElement.random();
    }

    // === RENDER ===

    private void draw() {
        ctx.clearRect(0, 0, canvas.getWidth(), canvas.getHeight());
        ctx.setFillStyle("#f5f0e0");
        ctx.fillRect(0, 0, canvas.getWidth(), canvas.getHeight());

        int total = cellSize + padding;
        for (int r = 0; r < gridSize; r++) {
            for (int c = 0; c < gridSize; c++) {
                int x = padding + c * total;
                int y = padding + r * total;

                ctx.setFillStyle((r + c) % 2 == 0 ? "#e8e0cc" : "#ddd5c0");
                ctx.fillRect(x, y, cellSize, cellSize);

                if (r == selectedRow && c == selectedCol) {
                    ctx.setStrokeStyle("#FFD700");
                    ctx.setLineWidth(3);
                    ctx.strokeRect(x + 2, y + 2, cellSize - 4, cellSize - 4);
                }

                if (grid[r][c] != null) {
                    GameElement el = grid[r][c];
                    if (el.isTarget) {
                        ctx.setFillStyle("rgba(255, 215, 0, 0.15)");
                        ctx.fillRect(x, y, cellSize, cellSize);
                    }
                    ctx.setFont((int)(cellSize * 0.7) + "px Arial");
                    ctx.setTextAlign("center");
                    ctx.setTextBaseline("middle");
                    ctx.fillText(el.icon, x + cellSize / 2.0, y + cellSize / 2.0);
                }
            }
        }
    }

    // === UI ===

    private void updateUI() {
        UIHelper.setText("dig-score", "Очки: " + score);
        UIHelper.setText("dig-target", "Цель: " + targetScore);
        UIHelper.setText("dig-moves", "Ходы: " + moves);
        UIHelper.setText("dig-size", "Поле: " + gridSize + "×" + gridSize);
        UIHelper.setText("dig-bones", "🦴 " + bonesCollected);
    }

    private void checkState() {
        if (score >= targetScore) {
            win();
        } else if (moves <= 0) {
            lose();
        }
    }

    private void win() {
        playing = false;
        showMessage("🎉 Отлично! " + score + " очков!", true);
        Window.setTimeout(() -> onCompleteCallback.call(score), 1500);
    }

    private void lose() {
        playing = false;
        showMessage("😔 Не хватило ходов!", false);
        Window.setTimeout(() -> onFailCallback.run(), 1500);
    }

    private void showMessage(String text, boolean success) {
        HTMLElement el = UIHelper.byId("dig-message");
        if (el == null) return;
        el.withText(text);
        UIHelper.setStyle(el, "display", "block");
        UIHelper.setStyle(el, "color", success ? "#27ae60" : "#e74c3c");
        Window.setTimeout(() -> UIHelper.setStyle(el, "display", "none"), 1500);
    }
}
